source/common/stats/char_star_set.cc

```diff
@@ -0,0 +1,49 @@
+#include "common/stats/char_star_set.h"
+
+#include "common/common/utility.h"
+
+namespace Envoy {
+namespace Stats {
+
+CharStarSet::~CharStarSet() {
+  // It's generally safer to clear associative containers before mutating
+  // elements of them (e.g. by deleting them), unless you know the container
+  // internals well, 2-phase deletes like are more obviously correct.
+  std::vector<char*> keys;
+  keys.reserve(hash_set_.size());
+  for (char* p : hash_set_) {
+    keys.push_back(p);
+  }
+
+  hash_set_.clear();
+  for (char* p : keys) {
+    delete[] p;
+  }
+}
+
+const char* CharStarSet::insert(absl::string_view str) {
+  size_t bytes = str.size() + 1;
+  char* ptr = new char[bytes];
+  StringUtil::strlcpy(ptr, str.data(), bytes);
+  auto insertion = hash_set_.insert(ptr);
+  if (!insertion.second) {
+    delete[] ptr;
```

Comment:
Couldn't we have the hash key be a `unique_ptr<char[]>` which would greatly simplify a bunch of the memory management in this class?

Reply:
I think that making the set key be a std::unique_ptr<char[]> would simplify some things but would complicate the implementation of find(), which would need to construct a unique_ptr object from a const char* it doesn't own in order to do the set lookup. That seems worse than the current state.

However I did change the destructor and insert() to use a std::unique_ptr as a temp, and I think that does help a little -- reduces lines of code anyway. PTAL. I could go either way. Or if you have a more specific suggestion about how to implement this interface more cleanly I'm fine with that too.